### ISSUE TITLE
Instances created with empty hostname fail to boot

### ISSUE THREAD

**Issue by taspelund** [oxidecomputer/omicron#4923]
Instances consistently fail to boot if the hostname is not set during instance creation.  This boot failure occurs for instances with or without a disk attached (no disk = firmware/pxe screen on serial console).  This issue is not observed for instances created via the console, because it copies `name` into `hostname` when the hostname isn't explicitly defined.

This was encountered on dogfood running a release somewhere between v5 and the upcoming v6 rc.

Instances that never start eventually hit the 60s timeout for the instance-create saga and transition to `failed` state:
```
20:54:13.401Z INFO 65a11c18-7f59-41ac-b9e7-680627f996e7 (ServerContext): start saga: ensuring instance is registered on sled
    file = nexus/src/app/sagas/instance_start.rs:512
    instance_id = a4badf7f-dcfc-4558-9d91-90bfcffdb834
    saga_id = 96427dbf-7b04-4223-b4c6-04958d7fdf1f
    saga_name = instance-start
    sled_id = 2707b587-9c7f-4fb0-a7af-37c3b7a9a0fa
20:55:13.661Z ERRO 65a11c18-7f59-41ac-b9e7-680627f996e7 (ServerContext): marking instance failed due to sled agent API error
    error = SledAgentInstancePutError(Communication Error: error sending request for url (http://[fd00:1122:3344:103::1]:12345/instances/a4badf7f-dcfc-4558-9d91-90bfcffdb834): operation timed out)
    file = nexus/src/app/instance.rs:1275
    instance_id = a4badf7f-dcfc-4558-9d91-90bfcffdb834
```

Creation of broken instance:
```
➜  oxide-aarch64-apple-darwin ./oxide instance create --name empty --description "" --hostname "" --memory 8589934592 --ncpus 2 --start false --project trey-trey-trey
success
Instance {
    description: "",
    hostname: "",
    id: 58c51ef7-a8cd-4b02-bf0d-c9f09aca9706,
    memory: ByteCount(
        8589934592,
    ),
    name: Name(
        "empty",
    ),
    ncpus: InstanceCpuCount(
        2,
    ),
    project_id: 5c3125f9-5aa8-455e-bc34-7c7576e2bf1c,
    run_state: Stopped,
    time_created: 2024-01-29T22:05:05.897656Z,
    time_modified: 2024-01-29T22:05:05.897656Z,
    time_run_state_updated: 2024-01-29T22:05:05.897656Z,
}
```

Creation of working instance:
```
➜  oxide-aarch64-apple-darwin ./oxide instance create --name not-empty --description "" --hostname "a" --memory 8589934592 --ncpus 2 --start false --project trey-trey-trey
success
Instance {
    description: "",
    hostname: "a",
    id: 77dc0a15-1721-48a5-b83c-05a1e5f5703b,
    memory: ByteCount(
        8589934592,
    ),
    name: Name(
        "not-empty",
    ),
    ncpus: InstanceCpuCount(
        2,
    ),
    project_id: 5c3125f9-5aa8-455e-bc34-7c7576e2bf1c,
    run_state: Stopped,
    time_created: 2024-01-29T22:05:16.149637Z,
    time_modified: 2024-01-29T22:05:16.149637Z,
    time_run_state_updated: 2024-01-29T22:05:16.149637Z,
}
```

Note that the only differences in these commands are `--name` and `--hostname`:
1. `./oxide instance create --name empty --description "" --hostname "" --memory 8589934592 --ncpus 2 --start false --project trey-trey-trey`
2. `./oxide instance create --name not-empty --description "" --hostname "a" --memory 8589934592 --ncpus 2 --start false --project trey-trey-trey`

I've tested multiple variations (even with the same `--name`) and found that using `--hostname ""` will result in an instance that cannot be started.

(thanks to @gjcolombo for all his help)

**Comment by bnaecker** [comment]
This is the original cause of #4931. We provide an invalid hostname from Nexus to the sled-agent, which causes it to deadlock as laid out in that issue. We should consider doing more validation of the hostname in Nexus, in addition to the work needed to prevent deadlocks to resolve #4931. We currently have a `NOT NULL` constraint on the hostname inside the DB schema, but nothing beyond that so far as I can see. There is [currently a TODO](https://github.com/oxidecomputer/omicron/blob/62547d22e74c4bb134b8d65e373ace793de7529f/nexus/types/src/external_api/params.rs#L1007) about making this a different type, which could presumably include validating the name as an RFC 1035 compliant name.

**Comment by bnaecker** [comment]
Closed by #4938
